### PR TITLE
added lesson materials for Intermediate Topics in Python

### PR DESCRIPTION
**Pull request by aays**
Contains .ipynb notebook as well as sample data for lesson (scores.txt)

### CODE REVIEW

**Comment by aays** [comment]
(Please don't be concerned by the large number of commits there - I swear I'll understand how to use Git properly one of these days)

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: `isinstance(num, str)`

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: Why do we want to skip the 89% we added?

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: `isinstance(num, str)`

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: `isinstance(num, str)`

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: What is "this"? `pass` is not really related to loops, specifically.

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: Doesn't appear that you have any use for `i`; why not `for item in mylist:`?

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: Ditto.

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: Ditto.

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: `instance(mylist[i], int)`

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: `isinstance(mylist[i], str)` (maybe also `elif`).

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: Could be `for i, item in enumerate(mylist):`; don't know when to mention that.

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: This logical expression is incorrect; you're asserting that `type(x)` is truthy and `type(y) != str`, not that both types aren't `str`. Should be `assert not isinstance(x, str) and not isinstance(y, str)` (parentheses unnecessary.)

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: Unnecessary parentheses.

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: Make into heading?

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: `i` is not needed; `for item in newlist:`

File: lessons/python/inttopics/inttopics.ipynb
Comment (QuLogic) [inline lessons/python/inttopics/inttopics.ipynb]: `i` is not needed; `for item in newlist:`

**Comment by QuLogic** [comment]
I think you'll still need to add a `lesson.md` even if it's minimal and just links to the notebook, or else this won't be showing up.

File: lessons/python/inttopics/inttopics.ipynb
Comment (joelostblom) [inline lessons/python/inttopics/inttopics.ipynb]: There is also the `extend` method, which gives what is often the expected behaviour when adding multiple items. It might be the preferred default choice here as well.

```python
# These two are the same
scores.extend([89, 90])
scores + [89, 90]

# And these two are the same
scores.append([89, 90])
scores + [[89, 90]]
```

File: lessons/python/inttopics/inttopics.ipynb
Comment (joelostblom) [inline lessons/python/inttopics/inttopics.ipynb]: This is not special behavior for list comprehensions, it happens just because we assigned the output to the variable `scores`.

I would actually turn this argument around and caution people against using the `append` or `extend` methods, since these modify the object in place. With list comprehension and simple list addition, you have the opportunity to review your new list by not assigning it to a variable and just having it print to stdout, until you arrive at the desired output.

File: lessons/python/inttopics/inttopics.ipynb
Comment (joelostblom) [inline lessons/python/inttopics/inttopics.ipynb]: Do you think it is worthwhile mentioning how to use [conditional expressions](https://docs.python.org/3/reference/expressions.html#conditional-expressions) in list comprehensions? Although this is not technically part of the comprehension syntax, it is the answer to "How do I use `else` in a list comprehension?". It might add some confusion since the syntax is slightly different, but it is good to know.

For your example, you could include the following to show how to keep both integers and strings while converting the strings to integers.

```python
[int(num)*0.4 if type(num) is str else num for num in scores]
```

**Review by joelostblom — CHANGES_REQUESTED** [review]
List compressions are the best ( ͡ᵔ ͜ʖ ͡ᵔ )

**Comment by joelostblom** [comment]
@aays Maybe it would have been better to submit that as a comment instead of requesting changes... My bad, still learning GitHub! Not sure If I am blocking something now, but I'll make sure to approve quickly when you make updates.

File: lessons/python/inttopics/inttopics.ipynb
Comment (aays) [reply lessons/python/inttopics/inttopics.ipynb]: That's a fair point - I was indeed referring to `pass` with 'this', but I'll clarify that it's not restricted to just loops.

File: lessons/python/inttopics/inttopics.ipynb
Comment (aays) [reply lessons/python/inttopics/inttopics.ipynb]: Bad habit! I'll correct that asap.

File: lessons/python/inttopics/inttopics.ipynb
Comment (aays) [reply lessons/python/inttopics/inttopics.ipynb]: I wanted to make a point just below this about how list comprehensions do not alter an object inplace like `append` does, but I suppose this is a bit of a clumsy way of leading to it. Should I just get rid of the else clause entirely?

File: lessons/python/inttopics/inttopics.ipynb
Comment (aays) [reply lessons/python/inttopics/inttopics.ipynb]: Eep! No idea how I bungled that. Fixing now.

File: lessons/python/inttopics/inttopics.ipynb
Comment (aays) [reply lessons/python/inttopics/inttopics.ipynb]: Was not aware of that! I've only ever added one thing at a time when appending, but that is handy to know. I'll update the notebook to reflect this as well.

File: lessons/python/inttopics/inttopics.ipynb
Comment (aays) [reply lessons/python/inttopics/inttopics.ipynb]: Nifty! I'll admit it's a little confusing at first glance, but it can't hurt to put it in there for the sake of completeness. Thanks for the tip!

**Review by joelostblom — APPROVED** [review]

**Comment by joelostblom** [comment]
@aays Great, just one additional thing to be aware of (don't think yo need to add anything, but it might come up). If anyone who is coding along opens `scores.txt` and saves it via their text editor, they might have a newline appended to the file (default behavior on any text editor following POSIX standards, e.g. vim and gedit.). Then they would need to modify the reading of the file to

```python
file.read().rstrip('\n').split(',')
```

**Comment by aays** [comment]
@QuLogic I added a `lesson.md` file! Let me know if it looks alright.

@joelostblom Cheers for pointing that out - that may very well come up, so I appreciate the tip :)

**Comment by lwjohnst86** [comment]
This is probably a bit of a stickler requirement, but could you please rename the folder from `inttopics` to `intermediate`? And the ipython file rename as `lesson.ipynb`.

File: lessons/python/inttopics/lesson.md
Comment (lwjohnst86) [inline lessons/python/inttopics/lesson.md]: After renaming the folder to intermediate, you'll have to fix this URL as well

**Review by lwjohnst86 — CHANGES_REQUESTED** [review]

File: lessons/python/inttopics/lesson.md
Comment (QuLogic) [inline lessons/python/inttopics/lesson.md]: Jupyter Notebook

File: lessons/python/inttopics/lesson.md
Comment (lwjohnst86) [inline lessons/python/inttopics/lesson.md]: And rename link here as well to `intermediate`

File: lessons/python/inttopics/lesson.md
Comment (lwjohnst86) [inline lessons/python/inttopics/lesson.md]: ~`s/inttopics/intermediate`~ `s/inttopics.ipynb/lesson.ipynb`

**Comment by QuLogic** [comment]
@joelostblom We don't require approval for merging, so it doesn't really matter if you leave a comment or request changes, except for making things clearer for whoever comes along later.

File: lessons/python/inttopics/lesson.md
Comment (aays) [reply lessons/python/inttopics/lesson.md]: Good catch!

**Comment by aays** [comment]
@lwjohnst86 Alright, I *think* I've fixed the folder name + incorporated suggested edits to `lesson.md`!

File: lessons/python/intermediate/lesson.md
Comment (lwjohnst86) [inline lessons/python/intermediate/lesson.md]: `s/inttopics/intermediate`

**Review by lwjohnst86 — APPROVED** [review]

**Pull request merged**
